RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc

```diff
@@ -0,0 +1,611 @@
+              if (useMask)
+                mask[t_i] = 1;
+            }
+          }
+        }
+      }
+    }  // backward
+  }    // seeding collection loop
+}
+
+void LinkingAlgoByDirectionGeometric::dumpLinksFound(std::vector<std::vector<unsigned>> &resultCollection,
+                                                     const char *label) const {
+  if (!(LinkingAlgoBase::algo_verbosity_ > LinkingAlgoBase::Advanced))
+    return;
+
+  LogDebug("LinkingAlgoByDirectionGeometric") << "All links found - " << label << "\n";
+  LogDebug("LinkingAlgoByDirectionGeometric") << "(seed can either be a track or trackster depending on the step)\n";
+  for (unsigned i = 0; i < resultCollection.size(); ++i) {
+    LogDebug("LinkingAlgoByDirectionGeometric") << "seed " << i << " - tracksters : ";
+    const auto &links = resultCollection[i];
+    for (unsigned j = 0; j < links.size(); ++j) {
+      LogDebug("LinkingAlgoByDirectionGeometric") << j;
+    }
+    LogDebug("LinkingAlgoByDirectionGeometric") << "\n";
+  }
+}
+
+void LinkingAlgoByDirectionGeometric::buildLayers() {
+  // build disks at HGCal front & EM-Had interface for track propagation
+
+  float zVal = hgcons_->waferZ(1, true);
+  std::pair<double, double> rMinMax = hgcons_->rangeR(zVal, true);
+
+  float zVal_interface = rhtools_.getPositionLayer(rhtools_.lastLayerEE()).z();
+  std::pair<double, double> rMinMax_interface = hgcons_->rangeR(zVal_interface, true);
+
+  for (int iSide = 0; iSide < 2; ++iSide) {
+    float zSide = (iSide == 0) ? (-1. * zVal) : zVal;
+    firstDisk_[iSide] =
+        std::make_unique<GeomDet>(Disk::build(Disk::PositionType(0, 0, zSide),
+                                              Disk::RotationType(),
+                                              SimpleDiskBounds(rMinMax.first, rMinMax.second, zSide - 0.5, zSide + 0.5))
+                                      .get());
+
+    zSide = (iSide == 0) ? (-1. * zVal_interface) : zVal_interface;
+    interfaceDisk_[iSide] = std::make_unique<GeomDet>(
+        Disk::build(Disk::PositionType(0, 0, zSide),
+                    Disk::RotationType(),
+                    SimpleDiskBounds(rMinMax_interface.first, rMinMax_interface.second, zSide - 0.5, zSide + 0.5))
+            .get());
+  }
+}
+
+void LinkingAlgoByDirectionGeometric::linkTracksters(const edm::Handle<std::vector<reco::Track>> tkH,
+                                                     const edm::ValueMap<float> &tkTime,
+                                                     const edm::ValueMap<float> &tkTimeErr,
+                                                     const edm::ValueMap<float> &tkTimeQual,
+                                                     const std::vector<reco::Muon> &muons,
+                                                     const edm::Handle<std::vector<Trackster>> tsH,
+                                                     std::vector<TICLCandidate> &resultLinked) {
+  constexpr double mpion = 0.13957;
```

Comment:
it might be good to define the TICL pi mass in a constexpr somewhere, also replacing

```
RecoHGCal/TICL/plugins/TICLCandidateFromTrackstersProducer.cc:          constexpr double mpion2 = 0.13957 * 0.13957;
RecoHGCal/TICL/plugins/TracksterP4FromTrackAndPCA.cc:        constexpr double mpion2 = 0.13957 * 0.13957;
RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:  constexpr double mpion = 0.13957;
```